src/Symfony/Component/Messenger/RoutableMessageBus.php

```diff
@@ -40,29 +40,39 @@ public function dispatch($envelope, array $stamps = []): Envelope
             throw new InvalidArgumentException('Messages passed to RoutableMessageBus::dispatch() must be inside an Envelope');
         }
 
-        /** @var BusNameStamp|null $busNameStamp */
-        $busNameStamp = $envelope->last(BusNameStamp::class);
+        /** @var ReceivedStamp|null $receivedStamp */
+        $receivedStamp = $envelope->last(ReceivedStamp::class);
 
-        if (null === $busNameStamp) {
-            if (null === $this->fallbackBus) {
-                throw new InvalidArgumentException(sprintf('Envelope is missing a BusNameStamp and no fallback message bus is configured on RoutableMessageBus.'));
+        if (null === $receivedStamp) {
+            // The RoutableMessageBus is only used in Worker context where the ReceivedStamp is added. So this should not happen.
+            throw new InvalidArgumentException('Envelope is missing a ReceivedStamp.');
+        }
+
+        $transportName = $receivedStamp->getTransportName();
+
+        /** @var SentToFailureTransportStamp|null $sentToFailureStamp */
+        $sentToFailureStamp = $envelope->last(SentToFailureTransportStamp::class);
+        if (null !== $sentToFailureStamp) {
+            // if the message was received from the failure transport, mark the message as received from the original transport and use the bus based on it
+            // this guarantees the same behavior when consuming from the failure transport (directly or via messenger:failed:retry) as when originally received
+            $originalReceiver = $sentToFailureStamp->getOriginalReceiverName();
+            // in case the original receiver does not exist anymore, use the bus configured for failure transport
+            if ($this->busLocator->has($originalReceiver)) {
+                $transportName = $originalReceiver;
             }
 
-            return $this->fallbackBus->dispatch($envelope, $stamps);
+            $envelope = $envelope->with(new ReceivedStamp($originalReceiver));
```

Comment:
This logic *does* feel a bit unnatural here. For example, all the "sent to failure" stuff is done via an event listener, but then we wire this part of the logic right inside this class. It could *still* be done as an event listener (to look for the `SentToFailureTransportStamp` and then add the `ReceivedStamp`, couldn't it? I think the complexity is this doing this part of the process if it were in a different listener:

> // in case the original receiver does not exist anymore, use the bus configured for failure transport

But I'm still not convinced this is the right place.

Reply:
This is actually fixing a bug and a requirement for `messenger:failed:retry` to work with multiple buses.
If messages end in the failure transport because messages are missing the BusNameStamp (external) or because developers renamed the bus, then those messages currently cannot be consumed at all from the failure transport because they can never go to the right bus.
By doing these checks in the RoutableMessageBus we can use the configured bus of the original transport.